Library/OcAppleKernelLib/KextPatcher.c

```diff
@@ -364,6 +364,103 @@ PatcherApplyGenericPatch (
+  IN OUT PRELINKED_CONTEXT      *PrelinkedContext
+  )
+{
+  MACH_SEGMENT_COMMAND_ANY  *Segment;
+  VOID                      *KextData;
+  UINT64                    AddressMax;
+  UINT64                    VirtualAddress;
+  UINT64                    Size;
+  UINT64                    MaxSize;
+  UINT32                    KextCount;
+  UINT32                    Index;
+  UINT32                    Index2;
+  XML_NODE                  *KextPlist;
+  UINT32                    KextPlistCount;
+  CONST CHAR8               *KextPlistKey;
+  XML_NODE                  *KextPlistValue;
+  CONST CHAR8               *KextIdentifier;
+
+  ASSERT (Identifier       != NULL);
+  ASSERT (PatcherContext   != NULL);
+  ASSERT (PrelinkedContext != NULL);
+
+  //
+  // Zero out kext memory through PatcherContext->MachContext.
+  //
+  Segment    = NULL;
+  AddressMax = 0;
+  MaxSize    = 0;
+  while ((Segment = MachoGetNextSegment (&PatcherContext->MachContext, Segment)) != NULL) {
+    VirtualAddress = PatcherContext->Is32Bit ? Segment->Segment32.VirtualAddress : Segment->Segment64.VirtualAddress;
+    Size           = PatcherContext->Is32Bit ? Segment->Segment32.Size : Segment->Segment64.Size;
+
+    AddressMax = MAX (VirtualAddress + Size, AddressMax);
+    MaxSize    = AddressMax - PatcherContext->VirtualBase;
+  }
+
+  KextData = MachoGetFilePointerByAddress (
+    &PatcherContext->MachContext,
+    PatcherContext->VirtualBase,
+    NULL
+    );
+  if (KextData == NULL) {
+    return EFI_UNSUPPORTED;
+  }
+  DEBUG ((
+    DEBUG_INFO,
+    "OCAK: Excluding %a - VirtualBase %Lx, MaxSize %Lx\n",
+    Identifier,
+    PatcherContext->VirtualBase,
+    MaxSize
+    ));
+  ZeroMem (KextData, (UINTN) (AddressMax + MaxSize));
+
+  //
+  // Find kext info to be removed in prelinked context.
+  //
+  KextCount      = XmlNodeChildren (PrelinkedContext->KextList);
+  KextPlist      = NULL;
+  KextPlistKey   = NULL;
+  KextIdentifier = NULL;
+  for (Index = 0; Index < KextCount; ++Index) {
+    KextPlist = PlistNodeCast (XmlNodeChild (PrelinkedContext->KextList, Index), PLIST_NODE_TYPE_DICT);
+    if (KextPlist == NULL) {
+      continue;
+    }
+
+    KextPlistCount = XmlNodeChildren (KextPlist);
+    for (Index2 = 0; Index2 < KextPlistCount; ++Index2) {
+      KextPlistKey = PlistKeyValue (PlistDictChild (KextPlist, Index2, &KextPlistValue));
+      if (KextPlistKey == NULL) {
+        continue;
+      }
+
+      if (AsciiStrCmp (KextPlistKey, INFO_BUNDLE_IDENTIFIER_KEY) == 0) {
+        KextIdentifier = XmlNodeContent (KextPlistValue);
+        if (PlistNodeCast (KextPlistValue, PLIST_NODE_TYPE_STRING) == NULL || KextIdentifier == NULL) {
+          return EFI_NOT_FOUND;
+        }
+        if (AsciiStrCmp (KextIdentifier, Identifier) == 0) {
+          DEBUG ((DEBUG_INFO, "OCAK: Erasing %a from prelinked at index %u\n", Identifier, Index));
+          //
+          // Erase kext.
+          //
+          XmlNodeRemove (PrelinkedContext->KextList, KextPlist);
```

Comment:
We have the index in "Index", no? Can we not call XmlNodeRemoveByIndex() directly?

Reply:
Makes sense, fixed.